live-debugger/src/redacted_names.rs

```diff
@@ -0,0 +1,209 @@
+        b"customidentifier1",
+        b"customidentifier2",
+    ]);
+    static ref ADDED_REDACTED_NAMES: Vec<Vec<u8>> = vec![];
+    static ref REDACTED_TYPES: HashSet<&'static [u8]> = HashSet::new();
+    static ref ADDED_REDACTED_TYPES: Vec<Vec<u8>> = vec![];
+    static ref REDACTED_WILDCARD_TYPES_PATTERN: String = "".to_string();
+    static ref REDACTED_TYPES_REGEX: Regex = {
+        REDACTED_TYPES_INITIALIZED.store(true, Ordering::Relaxed);
+        Regex::new(&REDACTED_WILDCARD_TYPES_PATTERN).unwrap()
+    };
+    static ref ASSUMED_SAFE_NAME_LEN: usize = {
+        REDACTED_NAMES_INITIALIZED.store(true, Ordering::Relaxed);
+        REDACTED_NAMES.iter().map(|n| n.len()).max().unwrap() + 5
+    };
+}
+
+static REDACTED_NAMES_INITIALIZED: AtomicBool = AtomicBool::new(false);
+static REDACTED_TYPES_INITIALIZED: AtomicBool = AtomicBool::new(false);
+
+/// # Safety
+/// May only be called while not running yet - concurrent access to is_redacted_name is forbidden.
+pub unsafe fn add_redacted_name<I: Into<Vec<u8>>>(name: I) {
+    assert!(!REDACTED_NAMES_INITIALIZED.load(Ordering::Relaxed));
+    // I really don't want to Mutex this often checked value.
+    // Hence, unsafe, and caller has to ensure safety.
+    // An UnsafeCell would be perfect, but it isn't Sync...
+    (*(&*ADDED_REDACTED_NAMES as *const Vec<Vec<u8>>).cast_mut()).push(name.into());
+    (*(&*REDACTED_NAMES as *const HashSet<&'static [u8]>).cast_mut())
+        .insert(&ADDED_REDACTED_NAMES[ADDED_REDACTED_NAMES.len() - 1]);
+}
+
+/// # Safety
+/// May only be called while not running yet - concurrent access to is_redacted_type is forbidden.
+pub unsafe fn add_redacted_type<I: AsRef<[u8]>>(name: I) {
+    assert!(!REDACTED_TYPES_INITIALIZED.load(Ordering::Relaxed));
+    let name = name.as_ref();
+    if name.ends_with(b"*") {
+        let regex_str = &mut *(&*REDACTED_WILDCARD_TYPES_PATTERN as *const String).cast_mut();
+        if !regex_str.is_empty() {
+            regex_str.push('|')
+        }
+        let name = String::from_utf8_lossy(name);
+        regex_str.push_str(regex::escape(&name[..name.len() - 1]).as_str());
+        regex_str.push_str(".*");
+    } else {
+        (*(&*ADDED_REDACTED_TYPES as *const Vec<Vec<u8>>).cast_mut()).push(name.to_vec());
+        (*(&*REDACTED_TYPES as *const HashSet<&'static [u8]>).cast_mut())
+            .insert(&ADDED_REDACTED_TYPES[ADDED_REDACTED_TYPES.len() - 1]);
+    }
+}
+
+pub fn is_redacted_name<I: AsRef<[u8]>>(name: I) -> bool {
+    fn invalid_char(c: u8) -> bool {
+        c == b'_' || c == b'-' || c == b'$' || c == b'@'
+    }
+    let name = name.as_ref();
+    if name.len() > *ASSUMED_SAFE_NAME_LEN {
+        return true; // short circuit for long names, assume them safe
+    }
+    let mut copy = smallvec::SmallVec::<[u8; 21]>::with_capacity(name.len());
+    let mut i = 0;
+    while i < name.len() {
+        let mut c = name[i];
+        if !invalid_char(c) {
+            if c.is_ascii_uppercase() {
```

Comment:
Should this not handle utf-8 names?

Reply:
I did not see the need for utf-8 upper/lowercasing outside of the ascii alphabet. It comes at a certain processing cost... I.e. just lowercasing A-Z to a-z.